### DISCUSSION TITLE
Segmentation and Classification Task

### DISCUSSION THREAD

**Discussion by Meddebma** [Project-MONAI/MONAI#2486] [Q&A]
Hi, 
I successfully segmented spleens in my dataset using the following code:
https://github.com/Meddebma/pyradiomics/blob/master/3D_Segmentation_eval.ipynb

In a second step I tried to classify the segmentation results into two labels in this code:
https://github.com/Meddebma/pyradiomics/blob/master/3D_Classification_Dict.ipynb

unfortunately I've got an error concerning the segmentations which have for example the shape of (512, 512, 50, 2), how can I get rid of the last dimension(2) or how should I manipulate the data to do the classification task?

Thanks many times

**Comment by Nic-Ma** [comment]
Hi @Meddebma ,

Thanks for your interest here with MONAI APIs.
Sorry I don't quite understand your question. Do you mean to move your last dim(channel) to be first dim?
You can use `EnsureChannelFirst` transform or `AsChannelFirst` transform.

Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma,

Yes Maybe I can do that, but I dont get why there are two channels, can I take it as the first dim or should I remove the channel I dont need? and If I should how can I remove it,

Thanks

**Reply by Nic-Ma** [reply]
Hi @Meddebma ,

I think you can check the source image why it contains 2 channels.
Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma,

I think the source image contains 2 channels because of the UNET of the segmentation task, it provides 2 output channels, how can I get rid of one channel that I dont need?

**Reply by Nic-Ma** [reply]
Hi @Meddebma ,

I checked the notebook you provided:
```py
transforms=Compose([  ToTensor()])
check = ImageDataset(image_files=images, labels=labels,
                        transform=transforms)
check_loader = DataLoader(check, batch_size=1,
                          num_workers=1, pin_memory=torch.cuda.is_available())
for data,i in check_loader:
  inputs = data[0].to(device)
  #plt.imshow(data[0][:, :, 35], cmap="gray")
  print(inputs.shape)
```
I think **UNet** is not involved in this code, seems the source image data has 2 channels, right?

Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma, 

yes exactly, the source image data with 2 channels was the segmentation output result of the following notebook:
https://github.com/Meddebma/pyradiomics/blob/master/3D_Segmentation_eval.ipynb

**Reply by Nic-Ma** [reply]
So I think maybe you can change the UNet output to 1 channel, or use `SplitChannel` transform to split the data and only use 1 channel.

Thanks.

**Reply by Meddebma** [reply]
@Nic-Ma 
alright, what do I put in the channel_dim then?
SplitChanneld(keys=["img"],channel_dim =?)

**Reply by Meddebma** [reply]
@Nic-Ma 
I changed the segmentation model to 1 in the output channel 

UNet(
            dimensions=3,
            in_channels=1,
            out_channels=1,
            channels=(16, 32, 64, 128, 256),
            strides=(2, 2, 2, 2),
            num_res_units=2)

But I get 0 mean dice score in every training epoch

**Reply by Nic-Ma** [reply]
I think you also need to also change related logic, don't skip background anymore because you have 1 channel, etc.
@rijobro could you please help share some comments? I think you are expert on training tuning.

Thanks in advance.

**Reply by rijobro** [reply]
If 2-channel segmentation is working for you, then you can just stick to that (also the method used in our [spleen segmentation tutorial](https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/spleen_segmentation_3d.ipynb)). When we visualise in that tutorial, you can see we do `[0, 0, ...`, where the second 0 is getting the 0th channel (the spleen). So you just want to keep that channel and discard the other: `spleen_segmentations = spleen_segmenations[:, 0, ...]`.

If you've already saved your 2-channel images to file, that's OK, too. There might be a specific MONAI transform for removing channels, if not you can just do it with a lambda:

```python
Compose([
    LoadImage
    AsChannelFirst
    Lambda(lambda x: x[0]
)]
```

**Reply by Meddebma** [reply]
Hi @rijobro, 
I tried it with the dictionary based Transform: 
Lambdad(keys=["img"], func=lambda x: x[0]), 
I've got the error in this notebook: 
https://github.com/Meddebma/pyradiomics/blob/master/3D_Classification_Dict.ipynb

**Reply by rijobro** [reply]
I can't see how the error relates to your code. Try just doing something like `image = transforms(img_fnames[0])` to avoid the confusion of the dataloader and multiprocessing.

**Reply by rijobro** [reply]
Also just noticed you have an `AddChannel`, before the lambda function to remove all but the first channel. You need to use `AsChannelFirst` if your images have shape `[h,w,d,c]`.

**Reply by Meddebma** [reply]
Hi @rijobro ,
So I did :
train_files = [{"img": img, "label": label} for img, label in zip(images[:18], labels[:18])]
val_files = [{"img": img, "label": label} for img, label in zip(images[-18:], labels[-18:])]
transforms = Compose(
        [
            LoadImaged(keys=["img"]),
            AsChannelFirstd(keys=["img"]),
            Lambdad(keys=["img"], func=lambda x: x[0]),
            ToTensord(keys=["img"]),
        ]
    )
image = transforms(train_files[0])


the result was: 
`Modifying image pixdim from [0.78125 0.78125 5.      1.     ] to [  0.78125      0.78125      5.         347.64570884]
`
when I put all the data I get this error: 
`RuntimeError: DataLoader worker (pid 299) is killed by signal: Bus error. It is possible that dataloader's workers are out of shared memory. Please try to raise your shared memory limit.`

**Reply by rijobro** [reply]
I don't understand your first error message. It implies that voxel spacing is being changed, but none of your transforms should be doing that.

Can you add them in one-by-one, to see which is causing the error:

```python
transforms = Compose([
    LoadImaged(keys=["img"]),
    #AsChannelFirstd(keys=["img"]),
    #Lambdad(keys=["img"], func=lambda x: x[0]),
    #ToTensord(keys=["img"]),
]
)
image = transforms(train_files[0])
```

**Reply by Meddebma** [reply]
@rijobro 
it happens in the LoadImaged!

**Reply by rijobro** [reply]
Ok, you're going to have to make a minimum working example I'm afraid. As far as I can see you'd like to do segmentation then classification, but are stuck on reading the image back in at the start of the classification. So the minimum components you need are here:

```python
# imports

# create fake image

# create segmentation network

# put fake image through network

# save image to file

# load image back from file
```

**Reply by Meddebma** [reply]
@rijobro Well That is exactly what I tried to do with the segmentation notebook here:
https://github.com/Meddebma/pyradiomics/blob/master/3D_Segmentation_eval.ipynb
I even can open the segmentation files in an external Nifti Reader. So I dont get it why it doesnt work..

what do you mean with the fake image, where should I get this from?

**Reply by rijobro** [reply]
I can't see any errors in your most recent notebook.

By fake I mean simply create images, as I don't have access to your data. Can be done with `torch.rand``.

**Reply by Meddebma** [reply]
@rijobro alright I am working on that. 
I noticed that in the unet_training_dict tutorial the out_channels=1 where as in the lightning tutorial was 2. When I change the out_channels to 1 the training doesnt work(mean dice is 0 in every epoch). Do I have to change something different in addition to the out_channels?

https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/spleen_segmentation_3d_lightning.ipynb

I am asking because the segmentation training is much faster in the lighning notebook than in the unet_training_dict

**Reply by rijobro** [reply]
You can set your network to output either one or two segmentation classes, there are 3d examples doing both ([1 class](https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/unet_segmentation_3d_ignite.ipynb), [2 classes](https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/spleen_segmentation_3d_lightning.ipynb)). 

Pick one, and mirror the tutorial until you're happy with your results.

When you're happy, move on to saving your files. If you have 2 channels, you can either choose to save both channels or discard the background channel. If you only have 1 channel, you won't have to make this decision.

Then you can move onto classification. If you're reading in an image with 2 channels, again, you can optionally discard the background channel.

I think you need to build up slowly, making sure that you're happy with your results at each step. Currently I get the impression that you're changing both the segmentation and classification notebooks, and this is like trying to hit a moving target (your classification will need to change as the output of your segmentation changes). Hope this helps.

**Reply by Meddebma** [reply]
@rijobro 
thank you, 
I am very happy with the segmentation results of this notebook:
https://github.com/Meddebma/pyradiomics/blob/master/3D_Segmentation_eval.ipynb
can you help me how I can discard the background channel before saving? 
I am saving using the nifti saver

**Reply by rijobro** [reply]
Sure. I suspect that the following should simply work:

```python
saver.save_batch(val_outputs[:, 0, ...], val_data["image_meta_dict"])
```
i.e., all batches, 0th channel, all spatial dimensions.

Let me know if that doesn't work,

**Reply by Meddebma** [reply]
@rijobro,
I tried it, it took very long time writing the file but then the system crashed
https://github.com/Meddebma/pyradiomics/blob/master/3D_Segmentation_eval.ipynb

**Reply by rijobro** [reply]
can't see any errors

**Reply by Meddebma** [reply]
okey, and if I try to discard the background in the classification task,
where should I put
images = images[:, 0, ...]?

https://github.com/Meddebma/pyradiomics/blob/master/3D_Classification.ipynb

**Reply by rijobro** [reply]
if you want to remove the channel when reading them in, you'll need to do something like:
```python
t = Compose([
    LoadImage
    AsChannelFirst
    Lambda(lambda x: x[0])
```
Notice how during the read step, we use `[0]` rather than `[:, 0, ...]` because during the post-process we have the batch dimension, but during the reading, we don't.
